### ISSUE TITLE
Efficiently compressing vectors of numerical data

### ISSUE THREAD

**Issue by KrzysFR** [Cyan4973/zstd#134]
Continuing from my tests, I'm looking at the best way to compress vectors of numbers produced by various sensors.

My goal is to have the more "efficient" compression scheme, ie: not one that compresses the most, or is the fastest, but the one that gives the most bang for the buck.

**Test Results**

I'm testing multiple forms of encoding, shuffling, delta coding, and testing them with different block sizes, compression levels etc, then graphing the results to look for trends, or oddities (such as in #130).

Some sensors produce large integers values that infrequently change over time, ie they typically look like:
```
xs = [ 
  146554, 146561, 146566, 146566, 146566, 146566, 146566, 146566, 146566, 146566, 
  ..., // one week later
  242259, 242259, 242260, 242260, 242260, 242260, 242261, 242261, 242261, 242269,
  ..., // another week later
  290667, 290667, 290667, 290667, 290667, 290667, 290667, 290667, 290667, 290667
]
```
The full encoded vector is about 1.9 MB (4-byte elements).

For delta coding, the datasets becomes:
```
delta(xs) = [
  293108, 14, 10, 0, 0, 0, 0, 0, 0, 0,
  .., // week later
  2, 0, 2, 0, 0, 0, 2, 0, 0, 16,
  .., // week later
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0
]
// last - first = 144113
```

I'm anticipating very good compression ratio, but I'm currently only getting around 40-50x as-is (down to 40KB). This is pretty respectable, but I would have expected more than that given the nature of the data. Even when fiddling with various techniques like delta coding, byte shuffling, rle pre-filtering, the best I'm getting is about 70x on this specific dataset. (note: on other sets where the values did not move much, I'm getting way better compression ratios, but mostly due to delta+rle):

The compression ratio flattens almost immediately, ie there is not much gains in there (except spending more cpu time to remove on a few hundred bytes). That's why I'm only testing up to `zstd -10`.

Compression ratio: _higher is better_
![image](https://cloud.githubusercontent.com/assets/1662235/13280192/c8538d1c-dadc-11e5-89ed-d4d4e254ad19.png)

On the other hand, the compression time shows something interesting: _lower is better_
![image](https://cloud.githubusercontent.com/assets/1662235/13280233/ff871dd0-dadc-11e5-8061-625160e26390.png)

The blue/gray lines are all encodings where the values are stored as 4-bytes little-endian integers (ie: plain old int array, no shuffling). The lower colorful lines are all using shuffle+rle encoding (same as blosc).

There does seem to be an advantage to shuffling bytes around and then using a very fast rle encoding on that to reduce the size of the data to compress, yielding the same compression ratio but much faster.

Looking at both ratio and time separatly does not help much visualize this, so I'm using `(ratio - 1) / time_sec` as a comparison metric. I don't know if there is already a name for this unit of measure, but if not, I will call them "Pirate-Ninjas" (PNJ), i.e: "number of 'X' gained per seconds" (basically it's just <code>s<sup>-1</sup></code> but I like the fancy name better :sunglasses: )

Using this chart, I can now clearly see that there is something going on for very specific compression schemes, that are several orders of magnitude faster than others.

Compression effiency: _more pirates and more ninjas is always more fun!_
![image](https://cloud.githubusercontent.com/assets/1662235/13280354/c3c227d0-dadd-11e5-8fb2-8e74f1f8b509.png)

Same graph in log scale:
![image](https://cloud.githubusercontent.com/assets/1662235/13280387/fafdd5be-dadd-11e5-890e-0c62634665f1.png)

**Observations**

Again, all codecs using shuffling dominate the others by 1 or 2 orders of magnitude.

The current winners for this dataset are (faster -> slower)
- `int_shfl`: int32 values shuffled by blocks of 256KB, then a fast pass of RLE to reduce their size before compressing with lz4 or zstd 0 to 2. Most efficient (500K pirate ninjas!), compress by about x53.
- `int_delta_shfl`: same as above, but working on deltas instead of values. less "efficient" (300K pirate ninjas) even though it compressed by x70.
- `flt_shfl` and `flt_delta_shfl`: same as above but using 32-bit floats (with custom byte shuffling to have the 8-bit exponent in its own byte instead of shifted because of the sign bit). Down to the 100K pirate-ninja territory which is terrible.
- the rest is garbage (~10K PNJ or less)

For reference:
- absolute values: raw size is 1,973,764 bytes, down to 111,516 bytes after shuffling+rle.
- delta values: raw size is 1,973,764 bytes, down to 49,653 bytes after shuffling+rle.

We can see that part of the compression speed boost comes from compressing only 50-KB or 111KB instead of 1.9MB, but what is counterintuitive to me, is that compressing 111KB of absolute values is 2x more efficient than compressing 50KB of delta values (if we consider both ratio and times).

On others simpler data vectors, I'm getting hundreds of millions Pirate Ninjas, up to the current highscore of 1.2 Billions (1.2E9 PNJ) on a vector that has the same value. This means that, per CPU second, I can store 1.2 billions times the same vector in the same place that it takes in memory. Nice :)

**Questions**

I'm not exactly sure why there is such a different trend in compression time between shuffled and non-shuffled datasets. It does not come from long runs of 0s or other because they get erased during the RLE stage that happens BEFORE the compression.

While delta-coding gives the best compression ratio here (though with other datasets, it usually only gives me 3x-5x more than non-delta), it is not the most "efficient". Since delta-coding makes it more difficult to random access the vector, I'm not planing on using it (even though it compress a tiny bit more).

My current RLE implementation is very crude and naïve, so I could probably get even better speed with a more advanced one. Is there any way to use some internal methods already in zstd to perform this stage? I'm seeing some RLE related methods in `fse.c` but they seemed to be specialized.

__dataset for reference__ _(I wish github would allow me to past excel tables not as pictures but as pre-formatted tables :( )_
![image](https://cloud.githubusercontent.com/assets/1662235/13281443/323b097e-dae4-11e5-8157-ad213a7f9491.png)

**Comment by Cyan4973** [comment]
When you have a table of aligned elements, and especially when element's values share some locality (ie, value change just a little between 2 consecutive measurements), then using @FrancescAlted 's blosc is the way to go.

It will help both compression and speed, because higher order bytes will most likely remain identical across large sections of the table. Hence, they can be advantageously grouped and then compressed using a simple and fast RLE.

> I'm not exactly sure why there is such a different trend in compression time between shuffled and non-shuffled datasets. It does not come from long runs of 0s or other because they get erased during the RLE stage that happens BEFORE the compression.

The resulting content from RLE stage is likely much smaller after blosc than without. As a consequence, since amount of data to compress is smaller, 2nd compression stage ends up being relatively faster.

> While delta-coding gives the best compression ratio here (though with other datasets, it usually only gives me 3x-5x more than non-delta), it is not the most "efficient". 

This conclusion entirely depends on the metric selected to define "efficient". If it favors speed, as it seems in your example, then compression gains might not be large enough to compensate increased CPU time.

Should the metric weights more in favor of compression ratio (exemple : `R^3/T` instead of `R/T`), conclusion will likely  be impacted.

**Comment by KrzysFR** [comment]
I've tried comparing `R/T` and `R^3/T` and the results are not that much different, except that indeed the `int_delta_shuffle` (dark red) makes a come back, and becomes the winner for levels `3` to `10`. Though still 3x to 4x less "efficient" than `int_shuffle` (green) can be at level `1` or `2`, and without the need to allocate more memory when encoding/decoding (which is not measured in this test).

![image](https://cloud.githubusercontent.com/assets/1662235/13329596/13f472e0-dbf5-11e5-803e-2b2b04db2f48.png)
![image](https://cloud.githubusercontent.com/assets/1662235/13329601/19003918-dbf5-11e5-8a21-bf6391235721.png)

And I'm realizing than in most applications, you will probably have different scenarios:
- `R^3/T` when you have to transmit data over a flaky network connexion.
- `R/T` when storing the same data on a biffy database server where disk space is less an issue than cpu time...

This is not a simple problem to solve :)

**Comment by Cyan4973** [comment]
Indeed, which is why there are multiple compression levels available.

Selecting the "right" one is really application specific.

**Comment by FrancescAlted** [comment]
Cool experiments @KrzysFR .  And definitely, my experience is that for binary data there are a good deal of transformations out there that can help compressing better and specially faster.  My Blosc2 project is (among other things) an attempt to make it possible to combine several filters prior to the compression stage so that both processes happens in the same block pipeline (for better L1/L2 usage).

**Comment by KrzysFR** [comment]
Yes, I was inspired by your great work on Blosc2, and wanted to try it myself with concrete datasets, especially the filtering part.

Is there a tool/script that would run all combinations of codecs / filters possible, on a set of samples, that would produce a matrix of al the results in a way that can be easily charted? 

Right now I have a semi automated process for this, but it's a pain to copy/paste the data into excel to get nice graphs, and be able to see trends / make an informed decision on what works and what doesn't...

**Comment by FrancescAlted** [comment]
No, this part is missing.  With the [bench](
https://github.com/Blosc/c-blosc2/blob/master/bench/bench.c#L392) utility
you can test the different codecs in combination with shuffle or
bitshuffle, but that's all.  In Blosc2 you should be able to combine more
filters (including the new delta) in the same pipeline, but there is not an
utility to exercise that yet.

However, there is a simple example showing the combined effect of delta +
bitshuffle at
https://github.com/Blosc/c-blosc2/blob/master/examples/delta_schunk.c.  You
can take this as an starting point for a more complete utility.

2016-02-26 17:00 GMT+01:00 Christophe Chevalier <notifications@github.com>:

> Yes, I was inspired by your great work on Blosc2, and wanted to try it
> myself with concrete datasets, especially the filtering part.
>
> Is there a tool/script that would run all combinations of codecs / filters
> possible, on a set of samples, that would produce a matrix of al the
> results in a way that can be easily charted?
>
> Right now I have a semi automated process for this, but it's a pain to
> copy/paste the data into excel to get nice graphs, and be able to see
> trends / make an informed decision on what works and what doesn't...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Cyan4973/zstd/issues/134#issuecomment-189339754>.
>



-- 
Francesc Alted

**Comment by Cyan4973** [comment]
Is there anything left to do for this issue ?

My feeling is that this is a filter/preprocessing issue (blosc, etc,), not directly related to the compression algorithm itself.

What do you think ?

**Comment by KrzysFR** [comment]
I had to work on other issues and will probably only be able to resume work on this in a couple of months. I haven't re-run the tests with the most recent 0.6.x either but you seem to have done work mostly on higher levels, so the results above should be unchanged.

And yes it is definitely a filter/preprocessing job, though maybe in future versions if zstd would provide a very fast RLE or simple pre-filters (maybe using the hypothetical negative compression levels for this?) it could probably help a lot with the compressing exotic payloads, maybe as advanced API with helpers for 3D meshes, number vectors, and so on?

My current conclusion is that pre-processing combined with zstd levels 0, 1 or 2 is the current sweet spot for my use case.

Closing it for now until I have some time again.
